### PR TITLE
[V1] TP Ray executor

### PR DESCRIPTION
**Pull request by ruisearch42**
Support Ray Compiled Graphs based executor in V1

Perf results on L4 GPU: 
(Turned off prefix caching for both there is an issue in main branch which is being fixed right now).
Perf is at parity, Ray is about ~0.7% slower.

VLLM_USE_V1=1 python3 benchmarks/benchmark_latency.py --model meta-llama/Llama-3.1-8B-Instruct --tensor-parallel-size 2  --num-iters-warmup 5 --num-iters 20  --batch-size 8 --input-len 128 --output-len 256 --max-model-len 2048 --no-enable-prefix-caching --distributed-executor-backend ray
Avg latency: 9.42691584636923 seconds
10% percentile latency: 9.391981961019336 seconds
25% percentile latency: 9.410091992700472 seconds
50% percentile latency: 9.425441902829334 seconds
75% percentile latency: 9.445964987040497 seconds
90% percentile latency: 9.456209814222529 seconds
99% percentile latency: 9.459444067198783 seconds

VLLM_USE_V1=1 python3 benchmarks/benchmark_latency.py --model meta-llama/Llama-3.1-8B-Instruct --tensor-parallel-size 2  --num-iters-warmup 5 --num-iters 20  --batch-size 8 --input-len 128 --output-len 256 --max-model-len 2048 --no-enable-prefix-caching
Avg latency: 9.356813629437237 seconds
10% percentile latency: 9.326405790261925 seconds
25% percentile latency: 9.342471541371197 seconds
50% percentile latency: 9.360302247805521 seconds
75% percentile latency: 9.37355133111123 seconds
90% percentile latency: 9.380363538581879 seconds
99% percentile latency: 9.386704193898476 seconds

### CODE REVIEW

File: vllm/v1/executor/ray_executor.py
Comment (comaniac) [inline vllm/v1/executor/ray_executor.py:60]: In the single GPU case with num_gpus=0.9 for example, will `torch.cuda.mem_get_info()` only see 0.9 of the total memory? The reason for this question is in GPU worker we use the following logic to determine the KV blocks:

```python
_, total_gpu_memory = torch.cuda.mem_get_info()
...
available_kv_cache_memory = (
    total_gpu_memory * self.cache_config.gpu_memory_utilization -
    peak_memory)
```

So if `total_gpu_memory` is already 0.9, then we will only use 0.81 of the total memory.

File: vllm/v1/executor/ray_executor.py
Comment (comaniac) [inline vllm/v1/executor/ray_executor.py:323]: Could just be `logger.debug("Called check_health.")`

**Review by comaniac — COMMENTED** [review]
Overall LGTM

File: vllm/v1/executor/ray_executor.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_executor.py:60]: `num_gpus` should not affect the memory obtained from `torch.cuda.mem_get_info()`. For example, when running this script:

```
import ray

ray.init(num_gpus=1)

@ray.remote
class FractionalGPUActor:
    def ping(self):
        import torch
        gpu_id = ray.get_runtime_context().get_accelerator_ids()["GPU"]
        mem_info = torch.cuda.mem_get_info()
        print(f"gpu_id: {gpu_id}, mem_info: {mem_info}")

fractional_gpu_actors = [FractionalGPUActor.options(num_gpus=0.1*(i+1)).remote() for i in range(3)]
# Ray tries to pack GPUs if possible.
[ray.get(fractional_gpu_actors[i].ping.remote()) for i in range(3)]
```

We got:
```
(FractionalGPUActor pid=3762944) gpu_id: ['0'], mem_info: (23382261760, 23583784960)
(FractionalGPUActor pid=3762941) gpu_id: ['0'], mem_info: (23184015360, 23583784960)
(FractionalGPUActor pid=3762946) gpu_id: ['0'], mem_info: (22985703424, 23583784960)
```

The total memory is the same.

**Review by comaniac — APPROVED** [review]
LGTM. cc @WoosukKwon @njhill @tlrmchlsmth

**Comment by DarkLight1337** [comment]
Please merge from main to resolve the CI failure.

**Comment by WoosukKwon** [comment]
@tlrmchlsmth @youkaichao Can you please take a look?

**Comment by comaniac** [comment]
> we should wait for https://github.com/vllm-project/vllm/pull/11256 . I don't want to duplicate the code, large parts of the code can be shared.

I don't think we should be blocked by #11256 for the following reason:
1. #11256 touches a larger scope and seems not ready yet; while this PR is limited scoped and is ready. Even there might be temporary code duplication, we shouldn't let a PR blocks a ready-to-merge one.
2. This PR is for v1. Even it could reuse code from v0 after #11256 is merged, I don't see any problem merging this PR first. On the other hand, please feel free to remove duplicated code in #11256 after this PR.

File: vllm/v1/executor/ray_executor.py
Comment (tlrmchlsmth) [inline vllm/v1/executor/ray_executor.py:313]: Actually, only the worker with `rank==0` returns the output and the others return `None` in order to make the collective_rpc mechanism simpler and to avoid deserializing extra outputs. Could you fix up the comment accordingly?

**Review by tlrmchlsmth — COMMENTED** [review]
One question: Why does `execute_model` use `_compiled_ray_dag`, while the other IPC calls use `_run_workers`? Why not use `_compiled_ray_dag` everywhere?

**Comment by youkaichao** [comment]
@ruisearch42 can you explain the difference between the `vllm/v1/executor/ray_executor.py` and `vllm/executor/ray_gpu_executor.py` for the sake of review?

File: vllm/v1/executor/ray_executor.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_executor.py]: These are some stale comments that applied to the case that there is a driver worker as well. Removed as it does not apply here. Thanks for catching.

**Comment by ruisearch42** [comment]
> One question: Why does `execute_model` use `_compiled_ray_dag`, while the other IPC calls use `_run_workers`? Why not use `_compiled_ray_dag` everywhere?

Good question. `execute_model` is the method that is repeatedly called, so we set up a Ray Compiled Graph for repeated execution of the same functionality, which avoids recreating underlying Ray Objects for each execution. `_run_workers` is only used at initialization time (although called a few times), and there is not much benefit using the Ray Compiled Graph API, and we can simply use the Ray Core API.

**Comment by tlrmchlsmth** [comment]
> > One question: Why does `execute_model` use `_compiled_ray_dag`, while the other IPC calls use `_run_workers`? Why not use `_compiled_ray_dag` everywhere?
> 
> Good question. `execute_model` is the method that is repeatedly called, so we set up a Ray Compiled Graph for repeated execution of the same functionality, which avoids recreating underlying Ray Objects for each execution. `_run_workers` is only used at initialization time (although called a few times), and there is not much benefit using the Ray Compiled Graph API, and we can simply use the Ray Core API.

That makes sense, thanks. To clarify: can the ray compiled graph only be used for `execute_model`? I.E. when we add other hot path functions will they need to have their own ray compiled graphs?

**Comment by ruisearch42** [comment]
> @ruisearch42 can you explain the difference between the `vllm/v1/executor/ray_executor.py` and `vllm/executor/ray_gpu_executor.py` for the sake of review?

Hi yeah, the V1 executor only supports SPMD mode, and non-SPMD code path is cleaned up. Also, in future, the PP implementation will not use virtual engine, so the structuring/interface of the executor will be different.

**Comment by ruisearch42** [comment]
> > > One question: Why does `execute_model` use `_compiled_ray_dag`, while the other IPC calls use `_run_workers`? Why not use `_compiled_ray_dag` everywhere?
> > 
> > 
> > Good question. `execute_model` is the method that is repeatedly called, so we set up a Ray Compiled Graph for repeated execution of the same functionality, which avoids recreating underlying Ray Objects for each execution. `_run_workers` is only used at initialization time (although called a few times), and there is not much benefit using the Ray Compiled Graph API, and we can simply use the Ray Core API.
> 
> That makes sense, thanks. To clarify: can the ray compiled graph only be used for `execute_model`? I.E. when we add other hot path functions will they need to have their own ray compiled graphs?

Yes, ray compiled graph is only used for distributed model execution (`worker.execute_model`) so far and there is no plan to include other parts. So other functionality should be orthogonal.

File: tests/basic_correctness/test_basic_correctness.py
Comment (ruisearch42) [reply tests/basic_correctness/test_basic_correctness.py:133]: It does not work with ray executor. But I think it makes sense this option is for MP only as Ray is a different executor.

**Comment by tlrmchlsmth** [comment]
> Yes, ray compiled graph is only used for distributed model execution (`worker.execute_model`) so far and there is no plan to include other parts. So other functionality should be orthogonal.

What about other "hot path" operations like add/remove lora?

**Comment by ruisearch42** [comment]
> > Yes, ray compiled graph is only used for distributed model execution (`worker.execute_model`) so far and there is no plan to include other parts. So other functionality should be orthogonal.
> 
> What about other "hot path" operations like add/remove lora?

If the data size is large or requires GPU-GPU data transfer, then perhaps we can optimize with Compiled Graphs. But AFIAK it is not the case, so using the normal ray core API should be good.

File: tests/basic_correctness/test_basic_correctness.py
Comment (tlrmchlsmth) [reply tests/basic_correctness/test_basic_correctness.py:133]: I don't think this should have anything to do with the executor that's used. https://github.com/vllm-project/vllm/pull/9826 is the PR that introduced it and has a nice diagram of what's going on.

What happens if we try to run the Ray executor with the AsyncLLM?

File: vllm/v1/executor/ray_executor.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_executor.py]: Done.

File: vllm/v1/executor/ray_executor.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_executor.py]: Updated comment:
3. Finally, if the work is on a node with smaller IP address, it
                should be placed first. This is simply a tiebreaker to make
                sure the workers are sorted in a deterministic way.

File: vllm/v1/executor/ray_executor.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_executor.py]: Done.

File: vllm/v1/executor/ray_executor.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_executor.py:310]: This should return immediately.

File: vllm/v1/executor/ray_executor.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_executor.py:320]: Ray compiled graphs requires higher level ray versions, but V0 ray executor does not. We don't want to force all users to upgrade ray versions at this point. And if they do use compiled graphs, this error message should be clear for them to upgrade ray.

File: vllm/v1/executor/ray_executor.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_executor.py:299]: Compiled graph is used to trigger all worker model execution and getting results, which is very similar to other executors. It should be orthogonal to internals of model execution.

File: tests/basic_correctness/test_basic_correctness.py
Comment (ruisearch42) [reply tests/basic_correctness/test_basic_correctness.py:133]: When using `VLLM_ENABLE_V1_MULTIPROCESSING`, it actually caused ray worker initialization to hang, because of an uninitialized ray environment in the new process.

Comment (tlrmchlsmth) [reply tests/basic_correctness/test_basic_correctness.py:133]: We should fix this, since AsyncLLM allows detokenization to be overlapped with the forward pass, and is the default in V1 now.

We probably want to fix this by initializing the ray environment in the entry point for the process that owns the executor:
https://github.com/vllm-project/vllm/blob/48edab8041741a82a1fd2f4d463cc0f393561b05/vllm/v1/engine/core.py#L239-L276

I don't think this should be a blocker for this PR but could you look into fixing this in a follow-up soon?

Comment (ruisearch42) [reply tests/basic_correctness/test_basic_correctness.py:133]: Thanks, sounds good. I will follow up soon.

File: vllm/v1/executor/ray_utils.py
Comment (ruisearch42) [reply vllm/v1/executor/ray_utils.py:20]: There are some incompatibilities. For example, V0 WorkerWrapperBase assumes V0 worker (`WorkerBase`), but in V1 we use V1 workers. If we do that, then there are more if-else cases introduced, making both less readable.

Comment (tlrmchlsmth) [reply vllm/v1/executor/ray_utils.py:20]: In the V1 multiproc executor, we're using `WorkerWrapperBase`, so it isn't V0 specific, it works for V1 as well
https://github.com/vllm-project/vllm/blob/d573aeadcc891976f09d6d50f1a4f98c8ff809aa/vllm/v1/executor/multiproc_executor.py#L235-L238

The way this works is in the CUDA platform code we check if the V1 env is set and pick the V1 worker if so:
https://github.com/vllm-project/vllm/blob/d573aeadcc891976f09d6d50f1a4f98c8ff809aa/vllm/platforms/cuda.py#L134-L138

Could you try doing this?

Comment (ruisearch42) [reply vllm/v1/executor/ray_utils.py]: Thanks, I think you are right. Fixed!

**Review by tlrmchlsmth — APPROVED** [review]
Nice, looks good to me!

**Comment by ruisearch42** [comment]
Hey @youkaichao , We plan to merge the PR soon, please take a look.

**Pull request merged**
